tachyon/math/matrix/matrix_operations.h

```diff
@@ -0,0 +1,160 @@
+#ifndef TACHYON_MATH_MATRIX_MATRIX_OPERATIONS_H_
+#define TACHYON_MATH_MATRIX_MATRIX_OPERATIONS_H_
+
+#include "tachyon/base/openmp_util.h"
+#include "tachyon/math/matrix/matrix_types.h"
+
+namespace tachyon::math {
+
+// The motivation of the variants of |MulMatVec()| is as follows:
+
+// NOTE(chokobole): Eigen matrix multiplication has a computational
+// overhead unlike naive matrix multiplication.
+//
+// Example: Multiplying a 2x2 matrix by a 2x1 vector:
+//
+// +----+----+   +----+
+// | m₀ | m₁ | * | v₀ |
+// +----+----+   +----+
+// | m₂ | m₃ |   | v₁ |
+// +----+----+   +----+
+//
+// The operations involved in this multiplication are as follows:
+//
+// 1 * 1
+// 1 * 1
+// m₀ * v₀
+// m₀v₀ + 0
+// m₂ * v₀
+// m₂v₀ + 0
+// m₁ * v₁
+// m₁v₁ + m₀v₀
+// m₃ * v₁
+// m₃v₁ + m₂v₀
+// m₁v₁ + m₀v₀ * 1
+// m₁v₁ + m₀v₀ + 0
+// m₃v₁ + m₂v₀ * 1
+// m₃v₁ + m₂v₀ + 0
+
+template <typename Derived, typename Derived2,
+          typename F = typename Derived::Scalar>
+math::Vector<F> MulMatVec(const Eigen::MatrixBase<Derived>& matrix,
+                          const Eigen::MatrixBase<Derived2>& vector) {
+  static_assert(std::is_same_v<F, typename Derived2::Scalar>);
+
+  math::Vector<F> ret = math::Vector<F>::Constant(vector.size(), F::Zero());
+  OPENMP_PARALLEL_FOR(Eigen::Index i = 0; i < matrix.rows(); ++i) {
+    for (Eigen::Index j = 0; j < matrix.cols(); ++j) {
+      ret[i] += matrix(i, j) * vector[j];
+    }
+  }
+  return ret;
+}
+
+template <typename Derived, typename Derived2,
+          typename F = typename Derived::Scalar>
+math::Vector<F> MulMatVecSerial(const Eigen::MatrixBase<Derived>& matrix,
+                                const Eigen::MatrixBase<Derived2>& vector) {
+  static_assert(std::is_same_v<F, typename Derived2::Scalar>);
+
+  math::Vector<F> ret = math::Vector<F>::Constant(vector.size(), F::Zero());
+  for (Eigen::Index i = 0; i < matrix.rows(); ++i) {
+    for (Eigen::Index j = 0; j < matrix.cols(); ++j) {
+      ret[i] += matrix(i, j) * vector[j];
+    }
+  }
+  return ret;
+}
+
+template <typename Derived, typename Derived2, enum Eigen::AccessorLevels Level,
+          typename F = typename Derived::Scalar>
+math::Vector<F> MulMatVec(
+    const Eigen::MatrixBase<Derived>& matrix,
+    const Eigen::DenseCoeffsBase<Derived2, Level>& vector) {
+  static_assert(std::is_same_v<F, typename Derived2::Scalar>);
+
+  math::Vector<F> ret = math::Vector<F>::Constant(vector.size(), F::Zero());
+  if (vector.rows() == 1) {
+    OPENMP_PARALLEL_FOR(Eigen::Index i = 0; i < matrix.rows(); ++i) {
+      for (Eigen::Index j = 0; j < matrix.cols(); ++j) {
+        ret[i] += matrix(i, j) * vector(0, j);
+      }
+    }
+  } else if (vector.cols() == 1) {
+    OPENMP_PARALLEL_FOR(Eigen::Index i = 0; i < matrix.rows(); ++i) {
+      for (Eigen::Index j = 0; j < matrix.cols(); ++j) {
+        ret[i] += matrix(i, j) * vector(j, 0);
+      }
+    }
```

Comment:
Looking at [the eigen document](https://eigen.tuxfamily.org/dox/group__TutorialMatrixArithmetic.html), there is a transpose function, so it seems like optimization is possible in this way.

See section `Transposition and conjugation` of the document above.

If there is no problem with this, it may be necessary to make more modifications to vector operations in this code file.

```suggestion
if (vector.rows() == 1) {
    OPENMP_PARALLEL_FOR(Eigen::Index i = 0; i < matrix.rows(); ++i) {
      ret[i] = matrix.row(i) * vector.transpose();
    }
  } else if (vector.cols() == 1) {
    OPENMP_PARALLEL_FOR(Eigen::Index i = 0; i < matrix.rows(); ++i) {
      ret[i] = matrix.row(i) * vector;
    }
  }
```

Reply:
The point of this is to avoid using naive operation because of the overhead made by Eigen, so I'll not apply it!